pkg/dynamicrp/datamodel/dynamicresource.go

```diff
@@ -32,3 +39,226 @@ type DynamicResource struct {
 	// Properties stores the properties of the resource being tracked.
 	Properties map[string]any `json:"properties"`
 }
+
+func (d *DynamicResource) Status() map[string]any {
+	if d.Properties == nil {
+		d.Properties = map[string]any{}
+	}
+
+	// We make the assumption that the status is a map[string]any.
+	// If users define the status as something other than a map[string]any, that just won't work.
+	//
+	// Therefore we overwrite it.
+	obj, ok := d.Properties["status"]
+	if !ok {
+		d.Properties["status"] = map[string]any{}
+		return map[string]any{}
+	}
+
+	status, ok := obj.(map[string]any)
+	if !ok {
+		d.Properties["status"] = map[string]any{}
+		return map[string]any{}
+	}
+
+	return status
+}
+
+// GetRecipe implements datamodel.RecipeDataModel.
+func (d *DynamicResource) GetRecipe() *portableresources.ResourceRecipe {
+	defaultRecipe := &portableresources.ResourceRecipe{Name: portableresources.DefaultRecipeName}
+
+	if d.Properties == nil {
+		return defaultRecipe
+	}
+
+	obj, ok := d.Properties["recipe"]
+	if !ok {
+		return defaultRecipe
+	}
+
+	recipe, ok := obj.(map[string]any)
+	if !ok {
+		return defaultRecipe
+	}
+
+	// This is the best we can do. We require all of the data we store to be JSON-marshallable,
+	// and the data should have already been validated when it was set.
+	bs, err := json.Marshal(recipe)
+	if err != nil {
+		panic("failed to marshal recipe: " + err.Error())
+	}
+
+	result := portableresources.ResourceRecipe{}
+	err = json.Unmarshal(bs, &result)
+	if err != nil {
+		panic("failed to unmarshal recipe: " + err.Error())
+	}
+
+	return &result
+}
+
+// SetRecipe implements datamodel.RecipeDataModel.
+func (d *DynamicResource) SetRecipe(recipe *portableresources.ResourceRecipe) {
+	if d.Properties == nil {
+		d.Properties = map[string]any{}
+	}
+
+	if recipe == nil {
+		d.Properties["recipe"] = map[string]any{}
+		return
+	}
+
+	// This is the best we can do. We designed the ResourceRecipe type to be JSON-marshallable.
+	bs, err := json.Marshal(recipe)
+	if err != nil {
+		panic("failed to marshal recipe: " + err.Error())
+	}
+
+	store := map[string]any{}
+	err = json.Unmarshal(bs, &store)
+	if err != nil {
+		panic("failed to unmarshal recipe: " + err.Error())
+	}
+
+	d.Properties["recipe"] = store
+}
+
+var _ rpv1.BasicResourcePropertiesAdapter = (*dynamicResourceBasicPropertiesAdapter)(nil)
+
+// ApplyDeploymentOutput implements v1.RadiusResourceModel.
+func (d *DynamicResource) ApplyDeploymentOutput(deploymentOutput rpv1.DeploymentOutput) error {
+	status := d.Status()
+
+	// This is the best we can do. We require all of the data we store to be JSON-marshallable.
+	bs, err := json.Marshal(deploymentOutput.DeployedOutputResources)
+	if err != nil {
+		panic("failed to marshal output resources: " + err.Error())
```

Comment:
Can you update this to an error instead of panic since this is on the request path? The function signature already returns an error, so shouldn't require refactoring consumers of this function.